### PR TITLE
Migrate from sbt to bazel

### LINKED ISSUE #613: Migrate this repo to Bazel and merge with bsp-generators
Currently most of the code in this repo is written in Scala and built with sbt. However sbt is hardly a polyglot build tool. In [this repo](https://github.com/agluszak/bsp-generators) we (JetBrains) have been writing two new library generators for Kotlin and Rust, both written in Kotlin and built with Bazel. They use a new, simpler IR. This however causes a split in the ecosystem. Every new feature added to the smithy meta-definition (traits, edge cases, types of data...) requires reimplementing the smithy-to-ir translation layer in two projects. I've been working on migrating the existing codegen to Bazel and partially to the new IR - [basically merging the two repos together](https://github.com/agluszak/bsp-generators/tree/scala-codegen). The only remaining (but pretty big) roadblock is the website generator, because it uses mdoc and mdoc seems to be deeply integrated with sbt (and supports only JVM languages).

### PR DESCRIPTION
**Pull request by agluszak**
Closes #613 

I also rewrote the Scala codegen in Kotlin, so from now on all generators will use the same IR.

### CODE REVIEW

**Comment by agluszak** [comment]
Cc @KatKlo

**Comment by adpi2** [comment]
I can start reviewing it this week, but I have almost no experience with Bazel so I will have more questions than useful comments.

**Comment by lefou** [comment]
I don't have profound Bazel experience so I can't say anything about the changes. Looks like the docs already mention how to build with `bazel`. I wonder why all those format changes to the generated source files need to be part of the tooling change. This really obfuscates the more essential change.

**Comment by agluszak** [comment]
> I can start reviewing it this week, but I have almost no experience with Bazel so I will have more questions than useful comments.

No problem, I'll do my best to answer them!

> I wonder why all those format changes to the generated source files need to be part of the tooling change. This really obfuscates the more essential change.

That's because bazel-super-formatter I used supports prettier (for the website) and java formatting, but doesn't support scalafmt. I managed to disable java formatting for now, but I'll have to think how to handle scalafmt.

**Comment by adpi2** [comment]
> I managed to disable java formatting for now, but I'll have to think how to handle scalafmt.

You can use coursier to download the scalafmt binary in the repo, and then commit it to the repo:
```
coursier bootstrap org.scalameta:scalafmt-cli_2.13:3.7.14 \
  -r sonatype:snapshots --main org.scalafmt.cli.Cli \
  --standalone \
  -o scalafmt
```

File: .bazelrc
Comment (agluszak) [reply .bazelrc:1]: Idk if it's relying on the security manager, but without this line java throws an error with jdk versions >= 14 IIRC

**Comment by agluszak** [comment]
> > I managed to disable java formatting for now, but I'll have to think how to handle scalafmt.
> 
> You can use coursier to download the scalafmt binary in the repo, and then commit it to the repo:
> 
> ```
> coursier bootstrap org.scalameta:scalafmt-cli_2.13:3.7.14 \
>   -r sonatype:snapshots --main org.scalafmt.cli.Cli \
>   --standalone \
>   -o scalafmt
> ```

No need to, I'll plug scalafmt into the existing formatting pipeline, but it's going to take some work. It's already exposed in `rules_scala`

File: .bazelrc
Comment (agluszak) [reply .bazelrc]: Looks like I managed to fix that and we can remove this line now

**Comment by agluszak** [comment]
@lefou all formatting changes should be gone now

**Comment by lefou** [comment]
@agluszak Thanks. I have no further comments. LGTM.

File: CONTRIBUTING.md
Comment (adpi2) [inline CONTRIBUTING.md:38]: Don't we want to figure this out before merging?

File: WORKSPACE
Comment (adpi2) [inline WORKSPACE:47]: The latest Scala 2.13 version is "2.13.12"

```suggestion
scala_config(scala_version = "2.13.6")
```

File: codegen/src/main/kotlin/org/jetbrains/bsp/generators/CodegenFile.kt
Comment (adpi2) [inline codegen/src/main/kotlin/org/jetbrains/bsp/generators/CodegenFile.kt:1]: What's the point of translating `codegen` to Kotlin? Why do we need to change the package name?

File: old-scala-codegen/BUILD
Comment (adpi2) [inline old-scala-codegen/BUILD:3]: So you split `codegen` in two? What's the difference between `codegen` and `old-scala-codegen\?

File: resources/buildServerProtocol.svg
Comment (adpi2) [inline resources/buildServerProtocol.svg:1]: Don't we need those images anymore?

File: spec-traits/src/main/kotlin/traits/DataKindTrait.kt
Comment (adpi2) [inline spec-traits/src/main/kotlin/traits/DataKindTrait.kt:13]: Here again how is it better to translate from Java to Kotlin?

File: website/pnpm-lock.yaml
Comment (adpi2) [inline website/pnpm-lock.yaml:1]: How is it better to use `pnpm` instead of `yarn`?

File: codegen/src/main/kotlin/org/jetbrains/bsp/generators/CodegenFile.kt
Comment (agluszak) [reply codegen/src/main/kotlin/org/jetbrains/bsp/generators/CodegenFile.kt:1]: Our new generators (for Kotlin and Rust) are written in Kotlin and use the `org.jetbrains.bsp` package name. I didn't include them here, but you can see them [here](https://github.com/agluszak/bsp-generators)

File: old-scala-codegen/BUILD
Comment (agluszak) [reply old-scala-codegen/BUILD:3]: old-scala-codegen is the codegen as it was previously in this repo. codegen is copied from bsp-generators

File: spec-traits/src/main/kotlin/traits/DataKindTrait.kt
Comment (agluszak) [reply spec-traits/src/main/kotlin/traits/DataKindTrait.kt:13]: it's much much shorter

File: website/pnpm-lock.yaml
Comment (agluszak) [reply website/pnpm-lock.yaml:1]: it's not better, but nodejs rules for bazel require using pnpm and don't support yarn

File: WORKSPACE
Comment (agluszak) [reply WORKSPACE:47]: When I was writing this `rules_scala` didn't support 2.13.12, but apparently[ the newest version released last week does](https://github.com/bazelbuild/rules_scala/releases/tag/v6.2.0)

File: CONTRIBUTING.md
Comment (agluszak) [reply CONTRIBUTING.md:38]: Yup, Danil Tereshchenko from our team is working on it

File: resources/buildServerProtocol.svg
Comment (jastice) [reply resources/buildServerProtocol.svg:1]: Aren't they used for the readme/website logo? Or are those others?

File: spec/src/main/resources/META-INF/smithy/manifest
Comment (KatKlo) [inline spec/src/main/resources/META-INF/smithy/manifest:13]: bsp/extensions/rust.smithy is doubled

File: codegen/src/main/kotlin/org/jetbrains/bsp/generators/ir/IrConfig.kt
Comment (agluszak) [inline codegen/src/main/kotlin/org/jetbrains/bsp/generators/ir/IrConfig.kt:52]: hmm, that's formatted incorrectly

**Comment by agluszak** [comment]
@adpi2 @lefou @jastice I figured out how to handle maven publishing, but I don't see any other way to test if it works with CI than actually running it...

**Comment by agluszak** [comment]
https://github.com/build-server-protocol/build-server-protocol/actions/runs/8192729876/job/22404911367 works!

**Comment by agluszak** [comment]
regarding 
image
don't worry, 99% of it the npm lockfile translated into bzlmod lockfile format which is... very verbose

File: bsp4j/BUILD
Comment (adpi2) [inline bsp4j/BUILD:1]: Why do `BUILD` files sometimes have a `.bazel` extension and sometimes not?

File: bsp4j/src/org/eclipse/lsp4j/util/Preconditions.java
Comment (adpi2) [inline bsp4j/src/org/eclipse/lsp4j/util/Preconditions.java:19]: Why don't we have a 4-space indentation in this Java file?

File: bsp4s/src/org/jetbrains/bsp/util/CustomCodec.scala
Comment (adpi2) [inline bsp4s/src/org/jetbrains/bsp/util/CustomCodec.scala:1]: Why did you change the package of this file? This is a breakage of the binary compatibility. Which means that in theory we should bump the major version of the library.

File: maven_install.json
Comment (adpi2) [inline maven_install.json:1]: What is this file?

File: smithy-build.json
Comment (adpi2) [inline smithy-build.json:1]: What was this file?

File: MODULE.bazel
Comment (adpi2) [inline MODULE.bazel:29]: If we don't compile to `2.12` anymore we should not need this `compat` dependency. We can try to remove it in this PR or another one.

File: codegen/src/main/kotlin/org/jetbrains/bsp/generators/dsl/dsl.kt
Comment (adpi2) [inline codegen/src/main/kotlin/org/jetbrains/bsp/generators/dsl/dsl.kt:1]: I suggest we use the same package for all the codegen code: `bsp.codegen`.

File: old-scala-codegen/BUILD
Comment (adpi2) [inline old-scala-codegen/BUILD:4]: Why is this `old`? It is still in use. Do you have a plan to replace it completely in the near future?

File: MODULE.bazel
Comment (adpi2) [reply MODULE.bazel:29]: Will Bazel be able to cross-compile to Scala 2.13 and Scala 3?

**Review by adpi2 — COMMENTED** [review]
This review contains mostly questions.

As I have very few knowledge about Bazel it would be good to have a review from someone else as well. Maybe @jastice?

File: old-scala-codegen/BUILD
Comment (agluszak) [reply old-scala-codegen/BUILD:4]: yes, that's the plan

File: MODULE.bazel
Comment (agluszak) [reply MODULE.bazel:32]: > Will Bazel be able to cross-compile to Scala 2.13 and Scala 3?
Yes, although [*will* is the key word here](https://github.com/bazelbuild/rules_scala/pull/1552). I mean, it should be possible now, but [it's cumbersome](https://eed3si9n.com/cross-build-anything-with-bazel/).

With the current setup we don't support Scala 3 anyway (or rather - support for it is not turned on by default, because I guess sbt supports cross building 2.13 and 3?)

File: smithy-build.json
Comment (agluszak) [reply smithy-build.json:1]: It was used by smithy CLI

File: maven_install.json
Comment (agluszak) [reply maven_install.json:1]: [it's a lock file ](https://github.com/bazelbuild/rules_jvm_external?tab=readme-ov-file#pinning-artifacts-and-integration-with-bazels-downloader)

File: CONTRIBUTING.md
Comment (agluszak) [reply CONTRIBUTING.md]: From the perspective of a (GitHub) user it works the same as before: you tag a release and the website is generated and deployed and artifacts are uploaded to maven.

The difference is that now we use https://github.com/JetBrains/bazel-sonatype instead of https://github.com/xerial/sbt-sonatype

File: .github/workflows/verify.yml
Comment (agluszak) [reply .github/workflows/verify.yml:59]: Done by `bazel run //:generate`

File: .github/workflows/verify.yml
Comment (agluszak) [reply .github/workflows/verify.yml:83]: Also done by `bazel run //:generate`

File: CONTRIBUTING.md
Comment (agluszak) [reply CONTRIBUTING.md:53]: Bazel is fully hermetic and installs dependencies inside its sandbox. This command installs them in the `node_modules` folder, where IDEs usually look for them.

File: MODULE.bazel
Comment (agluszak) [reply MODULE.bazel:1]: MODULE.bazel and WORKSPACE files are the entrypoints to a bazel workspace. Inside them you specify all rulesets/dependencies that will be used in your project. MODULE.bazel is the newer format, unfortunately not all dependencies have migrated to it ([`rules_scala` is the main culprit here... ](https://github.com/bazelbuild/rules_scala/issues/1482)).

BUILD.bazel files describe how to build specific targets.

File: bsp4j/BUILD
Comment (agluszak) [reply bsp4j/BUILD:1]: Fixed.

File: bsp4s/src/org/jetbrains/bsp/util/CustomCodec.scala
Comment (agluszak) [reply bsp4s/src/org/jetbrains/bsp/util/CustomCodec.scala:1]: accidentally...

File: CONTRIBUTING.md
Comment (adpi2) [reply CONTRIBUTING.md:34]: This is still unclear: Can I format the `codegen` code?

**Review by jastice — APPROVED** [review]
From my mostly surface-level review, LGTM! I can't evaluate the deeper logic without serious digging. Questions are only for clarification

File: .gitignore
Comment (agluszak) [reply .gitignore:56]: for now project views are editor specific tbh ;)

File: .github/workflows/verify.yml
Comment (agluszak) [reply .github/workflows/verify.yml:47]: why?

File: codegen/src/main/kotlin/bsp/codegen/common/FilesGenerator.kt
Comment (agluszak) [reply codegen/src/main/kotlin/bsp/codegen/common/FilesGenerator.kt:12]: because if there are no files to generate then probably there's been an error loading the schema

File: .github/workflows/verify.yml
Comment (lefou) [reply .github/workflows/verify.yml:47]: If they can be automatically generated, they should, unless there are good reasons not to (e.g. the tooling needs some remote resources temporarily unavailable or unstable).

**Pull request merged**
